### ISSUE TITLE
Ember Data 2.3 compatibility

### ISSUE THREAD

**Issue by devinus** [lytics/ember-data-model-fragments#174]
`DS` is no longer a global.

**Comment by devinus** [comment]
http://emberjs.com/blog/2016/01/12/ember-data-2-3-released.html

**Comment by slindberg** [comment]
Yep, this gonna be interesting.

**Comment by dwickern** [comment]
Does this mean you can remove the re-exported ember-data classes like `DS.Store`? Chrome opens your store.js instead of ember-data's which makes debugging it quite difficult.

**Comment by slindberg** [comment]
@dwickern that's bugged me for a long time too :stuck_out_tongue: The update will likely fix the issue.

**Comment by dfreeman** [comment]
There's also the fun fallout from Ember 2.3's `this.container` deprecation...

**Comment by htmldrum** [comment]
Any guidance for nubs who just end up crashing and burning on Uncaught ReferenceError: DS is not defined?

**Comment by kmiyashiro** [comment]
rollback to ember-data 2.2

**Comment by jakesjews** [comment]
We were able to upgrade by keeping ember-data in our bower.json file and setting it to 2.3.3 in both bower.json and package.json. You'll see a warning from ember-cli but you can ignore it.

**Comment by kmiyashiro** [comment]
@jakesjews did you check to see if ember-data is included twice in the build?

**Comment by jakesjews** [comment]
@kmiyashiro it looks like its only included one time. It looks like theres some checks inside ember-data for this case. https://github.com/emberjs/data/blob/f261ca8902219244994e79ba4785607d845d53ce/index.js#L68

**Comment by workmanw** [comment]
@slindberg et al. I would be willing to try to a PR for this. Steven it sounds like you have some domain knowledge here. If you can give me a little bit of direction, I'm happy to take it and try to run with it.

**Comment by dfreeman** [comment]
@slindberg @workmanw I'm also happy to put in some cycles if it helps move things forward here.

It looks like the build is currently pretty far off the ranch from the out-of-the-box addon setup (particularly with stubbing out the DS modules and the two-part npm/bower distribution) – is that in order to support producing a standalone artifact?

**Comment by slindberg** [comment]
@dfreeman It's mostly a hold-over from pre-addon days. It was the pattern a lot of early addons adopted, so I went with it. I will be very happy when the build artifacts are no longer a part of the repo (or the release process), and updating MF takes place only in `package.json`.

@workmanw @dfreeman Thanks for volunteering, I have extremely reduced bandwidth these days. ED 2.3 pulled the project into the ember-cli build, and MF should do the same. This will mostly be a file rename operation to move the `packages/` dir into `addon/`, in addition to fixing import paths to ED modules (adding the `-private` prefix). I'm not aware of any other code changes that would make MF incompatible.

Then it's just a matter of ripping out bower support. The downside to this is that backporting bugfixes is going to be a PITA, so ideally #173, #175, and #177 (maybe #176) get fixed first. Any interest in checking those out? :wink:

**Comment by workmanw** [comment]
@jakesjews Had started work on this here: https://github.com/eflexsystems/ember-data-model-fragments/tree/ember-cli-addon . I'm not sure what his status is at the moment.

I appreciate the fact that backporting fixes would be a pain, but I also think that Ember 2.3 compatible effects more users than those bugs do at the moment. My opinion would be to put in the cycles where ever we can.

**Comment by jakesjews** [comment]
@workmanw @slindberg I've reached a point with the Ember-CLI conversion where I think I need to add an ajax mocking library of some sort. Do either of you have any preferences for a good library to use?

I could use some help looking over the fork and making sure I'm not doing anything too crazy (I do need to fix the fact that all the tests are using moduleForAcceptance). Right now I have 311 of 341 assertions passing so hopefully its getting closer.

**Comment by dfreeman** [comment]
@jakesjews I'm a little surprised you need AJAX mocking, given that the existing tests didn't require it. What part of the changeover made that necessary?

**Comment by jakesjews** [comment]
@dfreeman To be honest I'm not 100% sure but if you look in the console its sending out a bunch of errors pertaining to ajax. Its possible it wasn't required before because of some custom setup that isn't happening now with the normal ember-cli tests. Would you be able to take a look at the fork to see if I'm missing something?

**Comment by jakesjews** [comment]
So I just did some digging and it turned out I needed to set `shouldBackgroundReloadRecord` to false on the dummy adapter.

**Comment by workmanw** [comment]
@jakesjews I don't have any preference. We use mockjax in our app ... but I've been hearing great things about ember-ajax (thought it might still be beta-ish). Doesn't matter to me :)

**Comment by jansegre** [comment]
I'm looking forward to this, I've started out with `ember-data` 2.4 (which [seem as compatible as 2.3](http://emberjs.com/blog/2016/02/29/ember-2-4-released.html)), this project would help a lot, but I'm not willing to pollute my project with temporary hacks to get it working.

I'm not experienced at all with ember addon development, but if there's anything I could do to help with this (docs for example) I'd be glad.

**Comment by workmanw** [comment]
@jansegre We are actively working on this. You can actually use this with Ember 2.4 now. There are just two caveats. 1) Is you will received deprecation warnings about usage of the container API. 2) You have to include ember-data in your bower.json. I personally don't consider either of those to be hacks.

**Deprecation warnings**
We have a PR that we're currently working to complete. Once that PR is complete we will cut a new version of this addon and this will resolve the aforementioned deprecation warnings..

**bower.json**
Once that new version is release, we will refactoring this addon to match the necessary structure outline by ember-cli-addons. @jakesjews has already done a large bulk of that work. Once this is complete we it will remove the need to include ember-data in your bower.json.

**Comment by jansegre** [comment]
@workmanw, great, I'll give it a shot then. I don't like the redundancy on __bower.json__ too much, but I can live with it for a while.

**Comment by bakura10** [comment]
Hi,

It does not seem to work at all with EmberJS 2.4 / Ember-Data 2.4 (using brain new Ember-CLI install). Here are the errors:

* Uncaught ReferenceError: DS is not defined
* Uncaught Error: Could not find module `ember-resolver` imported from `my-app/resolver`

What is the current workaround? :)

**Comment by kmiyashiro** [comment]
@bakura10 https://github.com/lytics/ember-data-model-fragments/issues/174#issuecomment-175809076

**Comment by bakura10** [comment]
Thanks, that works. However for those stumbling into this issue, since Ember Data 2.4 it's no longer requires to add it to bower.json and, acutally, when creating a new app with Ember-Cli the dependency in bower.json does not exist. To downgrade to ED 2.3 you must therefore manually add ED to the bower.json file.

**Comment by kernel-io** [comment]
How difficult would it be to add ED 2.4 support?

**Comment by workmanw** [comment]
@kernel-io We are actually using model-fragments with ember-data 2.4. AFAIK there are no code compatibility issues, it's just a matter of "add-on incompatibility" at the moment.

The workaround is to continue to include ember-data in your bower.json file ( https://github.com/lytics/ember-data-model-fragments/issues/174#issuecomment-175809076 ).

**Comment by workmanw** [comment]
I believe we can now close this issue. Ember-data 2.3+ is supported by ember-data-model-fragments 2.3.0. 👏 👏 👏 👏  🎉 

Thanks a lot to all for your patience, this was hopefully one of to more tricky upgrades we'll have in the 2.x cycle.

**Comment by slindberg** [comment]
I just want to give another huge thanks to @jakesjews and @workmanw for taking the lead on this update. Thank you!! ✨ 🐨 ✨
